cloudinit/net/network_state.py

```diff
@@ -526,21 +542,40 @@ def handle_bridge(self, command):
     def handle_infiniband(self, command):
         self.handle_physical(command)
 
-    @ensure_command_keys(['address'])
-    def handle_nameserver(self, command):
-        dns = self._network_state.get('dns')
+    def _parse_dns(self, command):
+        nameservers = []
+        search = []
         if 'address' in command:
             addrs = command['address']
             if not type(addrs) == list:
                 addrs = [addrs]
             for addr in addrs:
-                dns['nameservers'].append(addr)
+                nameservers.append(addr)
         if 'search' in command:
             paths = command['search']
             if not isinstance(paths, list):
                 paths = [paths]
             for path in paths:
-                dns['search'].append(path)
+                search.append(path)
+        return nameservers, search
+
+    @ensure_command_keys(['address'])
+    def handle_nameserver(self, command):
+        dns = self._network_state.get('dns')
+        nameservers, search = self._parse_dns(command)
+        if 'interface' in command:
+            self._interface_dns_map[command['interface']] = (
+                nameservers, search
+            )
+        else:
+            dns['nameservers'].extend(nameservers)
+            dns['search'].extend(search)
+
+    @ensure_command_keys(['address'])
+    def handle_individual_nameserver(self, command, iface):
```

Comment:
We might want to rename this because of ugly magic of CommandHandlerMeta which will attempt to map any 'handle_' method of NetworkStateInterpreter as a viable "command"
Note we've done the same with _handle_bond_bridge to avoid "registering" and handler for the top-level network config key "bond_bridge"

```suggestion
    def _handle_individual_nameserver(self, command, iface):
```